### DISCUSSION TITLE
Use multiple output interface.

### DISCUSSION THREAD

**Discussion by hosmelq** [chillerlan/php-qrcode#251] [Q&A] [answered]
I need to create QRs with gradients so I believe I need to use SVG but also I need to add a logo inside. Seeing the examples `imagickWithLogo` and `imagickConvertSVGtoPNG` they extend for different classes. I would like to know if it's possible to achieve this.

Thanks in advance.

**Comment by codemasher** [comment, marked as answer]
Do you have a specific requirement for the output (e.g. raster image)? If the QR Code is intended to be displayed on the web, there's no need to convert and you can just run with one of the SVG examples.
The `imagickWithLogo` is a pure ImageMagick example that is intended for raster output, while `imagickConvertSVGtoPNG` extends the SVG output class and adds conversion to a raster image on top. The SVG conversion with ImageMagick may or may not come with some issues (see #247) and it's recommended to use the Inkscape command line for conversion instead (as noted in the example).
In any case, there's no need to use multiple output classes.

**Reply by hosmelq** [reply]
I need to download the QRs in different formats JPG/PNG/SVG. I need to covert them mostly for the gradient requirement.

**Reply by codemasher** [reply]
In that case, your safest bet would be Inkscape, rather than ImageMagick.

**Reply by hosmelq** [reply]
Thank you very much, I will search for some examples. Do you know if I can use PNG logos with SVG? The example is with an SVG image but in my case users can only upload PNG/JPG.

**Reply by codemasher** [reply]
Yes, you can embed raster images in SVG with the `<image>`tag! See https://developer.mozilla.org/en-US/docs/Web/SVG/Element/image
The Inkscape command line documentation is over here: https://wiki.inkscape.org/wiki/Using_the_Command_Line
And you can even call it from within PHP via `exec()`: https://www.php.net/manual/en/function.exec or a command line library of your choice, e.g. https://github.com/symfony/process

**Reply by hosmelq** [reply]
Thank you for all the information. I will add the image as base64, I think adding it as a URL may introduce some problems is an image gets deleted.

**Reply by codemasher** [reply]
Yes that's generally a good idea, although it might bloat the SVG a bit. Good luck!
